[inductor] Clean up TRITON_CACHE_DIR

xw285cornell opened the pull request:
Summary:
As a follow up in https://github.com/pytorch/pytorch/pull/92664 (D42619405 (https://github.com/pytorch/pytorch/commit/e6a8267cf54af30e33de1ef22625e972afbf03ff)), clean up the TRITON_CACHE_DIR settings. There are a few places touching TRITON_CACHE_DIR:

1. triton/fb/triton_util.py: when import triton
2. caffe2/torch/_inductor/codecache.py
3. caffe2/torch/_inductor/triton_ops/autotune.py
4. triton/triton/python/triton/compiler.py

IIUC there are two entry points:
* kernel.run(args): 1 -> 3 -> 4
* async_compile(kernel): 1 -> 2 -> 3 -> 4
* calling triton jit-annoated func directly: 4

I'm removing the TRITON_CACHE_DIR in 1 and 2.

Test Plan: Run local repro

Differential Revision: D42694374



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel commented on torch/_inductor/triton_ops/autotune.py:
```diff
@@ -53,11 +54,14 @@ def __init__(self, fn, meta, configs, save_cache_hook, mutated_arg_names):
-        log.info(f"Triton cache directory: {triton_cache_dir}")
+        if os.getenv("TRITON_CACHE_DIR") is None:
+            os.environ["TRITON_CACHE_DIR"] = os.path.join(
+                tempfile.gettempdir(),
```
Let's make this relative to inductor's cache dir (`codecache.cache_dir()`)

jansel commented on torch/_inductor/triton_ops/autotune.py:
```diff
@@ -53,11 +54,14 @@ def __init__(self, fn, meta, configs, save_cache_hook, mutated_arg_names):
-        )
-        os.environ["TRITON_CACHE_DIR"] = triton_cache_dir
-        log.info(f"Triton cache directory: {triton_cache_dir}")
+        if os.getenv("TRITON_CACHE_DIR") is None:
```
Actually, this might be the wrong place to set this value -- there are some places where we don't actually use CachingAutotuner

bertmaher replied:
Oh this is a good point.  Do we always have inductor decorators (@pointwise, etc.) on the triton kernels we generate?  If not, they'll only go through Triton's tuner, right?

jansel replied:
Actually, that won't get it either.  The other place we call Triton is here:
https://github.com/pytorch/pytorch/blob/b399007a073d3a9ba4f6668c00d368fc9ee0f59a/torch/_inductor/select_algorithm.py#L383-L385

xw285cornell replied:
@jansel I think the main problem I had with PyCodeCache I may not be able to get device idx. But it seems I can call 

```
cuda_properties. current_device
```

So can we conclude that we can make it inside PyCodeCache?

xw285cornell replied:
@jansel I think another problem with setting this in TritonCodeCache or PyCodeCache is that it's a little higher level (called from getting fx graph). So if we already have a triton codegen it'll be called here. I think it's better to call it in lower level

xw285cornell commented:
Looks like failed test was disabled: https://github.com/pytorch/pytorch/pull/92945. Landing

xw285cornell commented:
@pytorchbot merge this please

xw285cornell commented:
@pytorchbot merge

xw285cornell committed:
[inductor] Clean up TRITON_CACHE_DIR (#92879)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/92879

As a follow up in https://github.com/pytorch/pytorch/pull/92664 (D42619405 (https://github.com/pytorch/pytorch/commit/e6a8267cf54af30e33de1ef22625e972afbf03ff)), clean up the TRITON_CACHE_DIR settings. There are a few places touching TRITON_CACHE_DIR:

1. triton/fb/triton_util.py: when import triton
2. caffe2/torch/_inductor/codecache.py
3. caffe2/torch/_inductor/triton_ops/autotune.py
4. triton/triton/python/triton/compiler.py

Since we decided to use device as the key, place 1 and 2 doesn't have device info so not good. We'll need to do at place 3. THere are places directly calling 4 w/o our autotune step which can still race.

IIUC there are two entry points:
* kernel.run(args): 1 -> 3 -> 4
* async_compile(kernel): 1 -> 2 -> 3 -> 4
* calling triton jit-annoated func directly: 4

I'm removing the TRITON_CACHE_DIR in 1 and 2.

#bypass-github-export-checks

Test Plan: Run local repro

Reviewed By: bertmaher, jansel

Differential Revision: D42694374

fbshipit-source-id: cf5b0cd36c5c47097050ec7453c543d4939e8dd2
